hail/src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -506,8 +509,20 @@ abstract class TableReader {
   def lowerGlobals(ctx: ExecuteContext, requestedGlobalsType: TStruct): IR =
     throw new LowererUnsupportedOperation(s"${ getClass.getSimpleName }.lowerGlobals not implemented")
 
-  def lower(ctx: ExecuteContext, requestedType: TableType): TableStage =
-    throw new LowererUnsupportedOperation(s"${ getClass.getSimpleName }.lower not implemented")
+  final def lower(ctx: ExecuteContext, requestedType: TableType, dropRows: Boolean): TableStage =
+    if (dropRows) {
+      val globals = lowerGlobals(ctx, requestedType.globalType)
+
+      TableStage(
+        globals,
+        RVDPartitioner.empty(ctx, requestedType.keyType),
+        TableStageDependency.none,
+        MakeStream(FastIndexedSeq(), TStream(TStruct.empty)),
+        (_: Ref) => MakeStream(FastIndexedSeq(), TStream(requestedType.rowType)))
+    } else
+      _lower(ctx, requestedType)
+
+  def _lower(ctx: ExecuteContext, requestedType: TableType): TableStage
```

Comment:
I'm not sure what this change adds other than a template pattern which I generally avoid whenever possible. I preferred keeping this logic in `LowerTableIr`.

How is `_lower` different to `lower`? `_lower` seems like something that shouldn't be public and calling it directly seems like a mistake.

Reply:
This is a pattern we use occasionally. `_lower` should really be protected, I'll make that change, and it is only called directly in `lower`, which is final. So `_lower` is the customization point for subclasses.

Reply:
What I was trying to understand is why do we need `_lower` at all when `TableRead` seems to be the only place that uses it?

Reply:
Ah, I understand now. That's fair. Addressed.